lib/linked_list.rb

```diff
@@ -18,71 +18,148 @@ def initialize
+      end
+      return current.data == value
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+      current = @head
+      max = @head.data
+      until current.next.nil?
+        current = current.next
+        if current.data > max
+          max = current.data
+        end
+      end
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def find_min
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+      current = @head
+      min = @head.data
+      until current.next.nil?
+        current = current.next
+        if current.data < min
+          min = current.data
+        end
+      end
+      return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      if @head.nil?
+        return 0
+      end
+      count = 1
+      current = @head
+      until current.next.nil?
+        current = current.next
+        count += 1
+      end
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+      current = @head
+      count = 0
+      until count == index || current.next.nil?
+        current = current.next
+        count += 1
+      end
+      return current.data
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+      current = @head
+      until current.nil?
+        print current.data
+        current = current.next
+      end
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def delete(value)
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      end
+      previous = @head
+      current = @head
+      until current.data == value || current.nil?
+        previous = current
+        current = current.next
+      end
+      previous.next == current.next

```

Comment:
You also need a check here to see if `value` is at the head of the list.

```suggestion
      if @head.nil?
        return nil
      elsif @head.data == value
        @head = @head.next
        return
      end
      previous = @head
      current = @head
      until current.data == value || current.nil?
        previous = current
        current = current.next
      end
      previous.next == current.next
```